datalad/core/distributed/push.py

```diff
@@ -188,6 +188,16 @@ def __call__(
             dataset, check_installed=True, purpose='pushing')
         ds_repo = ds.repo
 
+        if to and to not in ds_repo.get_remotes(
+                exclude_special_remotes=False):
+            # get again for proper error:
+            sr = ds_repo.get_remotes(exclude_special_remotes=False)
+            raise ValueError(
+                "Unknown push target '{}'. {}".format(
+                    to,
+                    'Known targets: {}.'.format(', '.join(repr(s) for s in sr))
+                    if sr
+                    else 'No targets configured in dataset.'))
```

Comment:
Failing earlier is of course good, but should this emit a `{"status": "error", ....}` result, as it did before?

Reply:
It would be nice to have a rule set to answer that question. I feel like I meander all the time. Here are the two thoughts that I use ATM:

- it happens on the dataset `push` is call on. A broken request, nothing can, was, or should be done: `ValueError`, because no action was performed 
- in the recursive case, a subdataset may not have a sibling defined, but publishing the dataset was attempted, and it failed: `error` result, because need to be able to continue (if desired)

Reply:
> It would be nice to have a rule set to answer that question. I feel like I meander all the time

Yeah, I agree that'd be nice.  It don't think I have anything like a clear rule set in my head, though I'd guess I lean towards results when I'm unsure (though looking at code might show that's not actually the case :).

> Here are the two thoughts that I use ATM: [...]

Thanks, that's helpful.  Even with your edited "What would not be [...]" point, I think raising a `ValueError` here does seem sensible then.

Reply:
It does, but you convinced me otherwise before ;-) The ability to force `push` through this kind of error could be useful. I'll keep it with an error result.